jdaviz/core/helpers.py

```diff
@@ -480,9 +492,26 @@ def _get_data(self, data_label=None, cls=None, subset_to_apply=None, function=No
                             warnings.warn(f"Not able to get {data_label} returned with"
                                           f" subset {subsets.label} applied of type {cls}."
                                           f" Exception: {e}")
+            elif (spatial_subset and spectral_subset and
+                  subsets.label.lower() == spectral_subset.lower()):
+                # Same logic as if statement
+                spec_sub = [spec_sub_data for spec_sub_data in subsets.subsets if
+                            spec_sub_data.data.label == data_label][0]
+        # Apply spectral subset to spatial subset if applicable
+        if spatial_subset and spectral_subset:
+            handler, _ = data_translator.get_handler_for(cls)
+            try:
+                spec_subset = handler.to_object(spec_sub, **object_kwargs)
+            except Exception as e:
+                warnings.warn(f"Not able to get {data_label} returned with"
+                              f" subset {spectral_subset} applied of type {cls}."
+                              f" Exception: {e}")
+            # Return collapsed Spectrum1D object with spectral subset mask applied
+            data.mask = ~spec_subset.mask
+
         return data
 
-    def get_data(self, data_label=None, cls=None, subset_to_apply=None):
+    def get_data(self, data_label=None, spatial_subset=None, spectral_subset=None, cls=None):
```

Comment:
should the default maybe not take any subset kwargs then and another implementation for imviz which _only_ takes `spatial_subset` and not `spectral_subset`?

Right now `imviz.get_data(data_label, spectral_subset='Subset 1')` works and just ignores the input `spectral_subset`.

Reply:
This is in the base core helper, so any config that inherits without overriding will see options for `spatial_subset` and `spectral_subset` in this method, which is not ideal.  If `get_data` does not take _any_ subsets by default, then the individual configs can handle their particular cases.  The private `_get_data` can still take both just to keep the code consolidated.

Reply:
Ok, are you satisfied with the addition to the imviz helper?

Reply:
Imviz does not accept spectral subsets either. So core get_data just returns data with no subsets applied? That's fine.

Reply:
But `get_data` is intended to be a user-facing method, and I worry it will cause confusion if it accepts things that will just be ignored or don't make sense for a particular config.  

For example, as-is in this PR, if I do `help(specviz2d.get_data)` it shows that it accepts both spatial and spectral subset.  If I create a _spectral_ subset called 'Subset 1' and call `specviz2d.get_data('Spectrum 2D', spatial_subset='Subset 1')`, it does not raise an error, but for some reason still applies that as a spectral mask?  If I create a second spectral subset and call `specviz2d.get_data('Spectrum 2D', spatial_subset='Subset 1', spectral_subset='Subset 2')` it seems to invert the mask of Subset 2... so its clearly not completely ignoring the passed value, but it is not acting in any sort of sensible way for specviz2d.

Reply:
It gets inverted here: https://github.com/spacetelescope/jdaviz/pull/2199/files#diff-cb5e0cef9bcbcb30559a6ce19d4bf05e855cfb36701fb325e690e53aeb0906eaR510 which I thought I needed to do since specutils and glue use different definitions of `mask=True`, that can be a simple change if you're sure it should be `data.mask = spec_subset.mask`.

As for making sure "Subset 1" is actually a spectral or spatial subset, is there already an API call that can check that? If not, I think I would have to call `get_subsets` and check based on that, which adds more time to the `get_data` call. 

Doing `help(specviz2d.get_data)` on the latest commit should not return that information now.

Reply:
But now `specviz2d.get_data` does not accept _any_ subsets.  I would expect it to accept `spectral_subset` but not `spatial_subset`.

Reply:
1. I think if its just an `isinstance` check on the subset, that should be pretty cheap and would be worth the check.  It would be nice to handle that consistently with #2207 (once revived), so could be pushed into that effort.  The tricky thing is that we don't forbid mixing spatial and spectral, but if you just look through the tree for the first entry and assume that applies to the whole subset, at least that will avoid the worst/easiest mistakes to make.
2. Please do double check, but I think that might have been caused by passing a spectral subset to both `spectral_subset` and `spatial_subset` (which is no longer possible in your recent changes).

Reply:
1. `isinstance` won't work for composite subsets though, right?
2. It looks correct to me.

Reply:
Sorry for missing this conversation earlier - in the demo earlier today it looked like the mask on the output `Spectrum1D` object was the opposite of what I would have expected. In `specutils`, `mask=True` is for points that are masked out, in other words not what we're interested in (so everything outside a subset). FWIW this is how it currently works in `get_spectra` on main:

![Screenshot 2023-05-22 at 12 56 09 PM](https://github.com/spacetelescope/jdaviz/assets/39831871/b3ef6e70-9ab4-4d3a-b570-0833bdd14842)

In that screenshot, I have a subset in the middle of the spectrum, so the start and end of the spectrum have `mask=True`. Now, do I remember off the top of my head if that is `~spec_subset.mask` or `spec_subset.mask`? Definitely not 😅

Reply:
I guess I can demo this at tag up but this is how `get_data` functions in this PR as well; the subset area is set with `mask[index]=False`. Is there another way to access the mask information in jdaviz that could conflict with this?

Reply:
Maybe I misunderstood what you showed at sprint review then 🤷 . If the area in the subset has `False` for the mask there then we're good.